File: haskell/proverb/src/Proverb.hs
module Proverb (recite) where

import Data.List (intercalate)
import Text.Printf (printf)

windows2 :: [x] -> [(x, x)]
windows2 [] = []
windows2 [_] = []
windows2 (a : b : rest) = (a, b) : windows2 (b : rest)

recite :: [String] -> String
recite [] = ""
recite xs = intercalate "\n" phrases
  where
    phrase = printf "For want of a %s the %s was lost."
    end = printf "And all for the want of a %s." $ head xs
    phrases = foldr (\(a, b) acc -> phrase a b : acc) [end] $ windows2 xs

File: zig/proverb/proverb.zig
const mem = @import("std").mem;
const fmt = @import("std").fmt;

pub fn recite(allocator: mem.Allocator, words: []const []const u8) (fmt.AllocPrintError || mem.Allocator.Error)![][]u8 {
    var proverb = try allocator.alloc([]u8, words.len);
    var idx: usize = 0;

    while (idx < words.len) : (idx += 1) {
        proverb[idx] = try if (idx == words.len - 1)
            fmt.allocPrint(
                allocator,
                "And all for the want of a {s}.\n",
                .{words[0]},
            )
        else
            fmt.allocPrint(
                allocator,
                "For want of a {s} the {s} was lost.\n",
                .{ words[idx], words[idx + 1] },
            );
    }

    return proverb;
}
